### PR TITLE
AF-1903: There is no way to upload multiple documents for Human Task forms

### PR DESCRIPTION
**Pull request by pefernan**
@tomasdavidorg @ederign could you please take a look?

@romartin I had to do some changes on some stunner form controls, would you mind take a look?

@tomasdavidorg I made some recent refactors and I need to add some tests... I'll be updating the PR soon. BTW you can use this project https://github.com/pefernan/dox as an example to configure the processes

This PR is part of an ensemble, please merge with:
https://github.com/kiegroup/kie-wb-common/pull/2580
https://github.com/kiegroup/kie-docs/pull/1551

### CODE REVIEW

**Comment by pefernan** [comment]
jenkins execute full downstream build please

**Comment by pefernan** [comment]
@tomasdavidorg I see jenkins failed, please use this business-central installation here https://mega.nz/#!ZFxy1ISZ!dL1N1cPAJxx3sRpsbMe698AvfzL-k2_7nW8pxwdn_tA to test while I fix it

**Comment by pefernan** [comment]
jenkins execute full downstream build please

**Comment by pefernan** [comment]
jenkins execute full downstream build please

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/test/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/definition/DocumentsListFieldDefinitionTest.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.workbench.common.forms.jbpm.model.authoring.documents.definition;
+
+import org.kie.workbench.common.forms.fields.shared.fieldTypes.AbstractFieldDefinitionTest;
+
+public class DocumentsListFieldDefinitionTest extends AbstractFieldDefinitionTest<DocumentListFieldDefinition> {
```
Comment (tomasdavidorg) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/test/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/definition/DocumentsListFieldDefinitionTest.java]: Are Empty and Full definitions the same?

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        this.dropFile(fileList[i]);
+    }
+};
+
+appformer.forms.Documents.prototype.dropFile = function (file) {
+    var id = Number(Math.random().toString().slice(2,11)).toString();
+
+    var document = new appformer.forms.Document(id, file.name, '', file.size, file.lastModified);
+
+    if(this.autoUpload) {
```
Comment (tomasdavidorg) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js]:
```suggestion
    if (this.autoUpload) {
```

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        var callback =  this.onDropCallback;
+        var fileReader = new FileReader();
+        fileReader.onload = function (event) {
+            if(event.target.readyState == FileReader.DONE) {
```
Comment (tomasdavidorg) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js]:
```suggestion
            if (event.target.readyState == FileReader.DONE) {
```

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if(!element) {
+        throw "Cannot bind documents upload to a null element";
+    }
+    if (!element.tagName) {
+        throw "Cannot bind documents upload to a non html element";
+    }
+
+    var tag = element.tagName.toUpperCase();
+
+    if(tag === "DIV") {
```
Comment (tomasdavidorg) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js]:
```suggestion
    if (tag === "DIV") {
```

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+appformer.forms.Documents.prototype.dropFilesList = function(fileList) {
+    for (var i = 0; i < fileList.length; i++) {
+        this.dropFile(fileList[i]);
+    }
+};
+
+appformer.forms.Documents.prototype.dropFile = function (file) {
+    var id = Number(Math.random().toString().slice(2,11)).toString();
```
Comment (tomasdavidorg) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js]:
```suggestion
    var id = Number(Math.random().toString().slice(2, 11)).toString();
```

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+};
+
+appformer.forms.Documents.get = function() {
+    return new appformer.forms.Documents(false);
+};
+
+appformer.forms.Documents.get = function(autoUpload) {
+    if(autoUpload == true) {
```
Comment (tomasdavidorg) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-client/src/main/resources/org/kie/workbench/common/forms/jbpm/client/resources/scripts/upload.js]:
```suggestion
    if (autoUpload == true) {
```

**Comment by tomasdavidorg** [comment]
I have finished code review... now I will test the war.

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/main/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/definition/DocumentListFieldDefinition.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+public class DocumentListFieldDefinition extends AbstractFieldDefinition {
+
+    public static final DocumentListFieldType FIELD_TYPE = new DocumentListFieldType();
+
+    @FormField(
+            labelKey = "maxDocuments",
+            helpMessageKey = "maxDocuments.helpMessage",
+            afterElement = "label"
+    )
+    private Integer maxDocuments = 0;
```
[the remark this answers is not in the document]
Comment (pefernan) [reply kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/main/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/definition/DocumentListFieldDefinition.java]: @tomasdavidorg yea, anything under 1 means no limit

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/main/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/provider/DocumentsListFieldProvider.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public DocumentListFieldDefinition getDefaultField() {
+        return new DocumentListFieldDefinition();
+    }
+
+    @Override
+    public int getPriority() {
+        return 16;
```
[the remark this answers is not in the document]
Comment (pefernan) [reply kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/main/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/provider/DocumentsListFieldProvider.java]: This priority is used to help sorting the field providers insided of the field manager... the main use is to make sure that the default field types from the form engine are being used over any custom field type might be added by another component (for example stunner is adding custom form controls).
This mechanism was added at the very beginning of the forms engine when we never and should be improved at some point, but for now does his part of the work :)

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/test/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/definition/DocumentsListFieldDefinitionTest.java (hunk 1 above)
Comment (pefernan) [reply kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-api/src/test/java/org/kie/workbench/common/forms/jbpm/model/authoring/documents/definition/DocumentsListFieldDefinitionTest.java]: Not really, the full should be a instance of the field definition with custom values, it is used to test the copy of the settings between two different instances.

thank you for pointing me this, I need to improve that test!

**Comment by tomasdavidorg** [comment]
There are few things that need to be resolved:
1. We have org.jbpm.document.Document and org.jbpm.document.Documents types which are hard to distinguish. Shouldn't be there any aliases for them?
2. The types need to be manually added into Process Variables Data Type. Shouldn't be offered by default as String or Boolean types? Screenshot from 2019-04-15 16-41-43
3. There is not way how to access multiple documents. They can be uploaded but they don't display in the Documents tab of process instance. https://issues.jboss.org/browse/JBPM-8421
4. Uploading of single document is confusing. When you select a file by Choose file button it is not uploaded but only selected. You have to click on the Upload button. There should be some warning that document is not yet uploaded or there should be only one button which directly uploads the document.

**Comment by pefernan** [comment]
@tomasdavidorg for #4 we already have this issue https://issues.jboss.org/browse/AF-1927
IMO this is not a blocker for this PR but we'll try to have it done soon if we have free time.

**Pull request merged**